### ISSUE TITLE
Storage Backend: Google Drive

### ISSUE THREAD

**Issue by Intensity** [restic/restic#527]
This is a feature request for Google Drive as a storage backend.

**Comment by fd0** [comment]
Thanks for the suggestion.

**Comment by middelink** [comment]
FYI
https://developers.google.com/drive/v3/web/quickstart/go
https://github.com/google/google-api-go-client/blob/master/drive/v3/drive-gen.go

**Comment by rxng** [comment]
any short term plans to implement this?

**Comment by fd0** [comment]
To be honest with you, there are no short-term plans for this. Sorry.

**Comment by xor-gate** [comment]
The people from the `rclone` tool have created an drive backend here: https://github.com/ncw/rclone/tree/master/drive

**Comment by naffit** [comment]
Hi, any possible news on this?
The latest Crashplan changes have made me look for an alternative and the unlimited space available when using G Suite for Business is really interesting - and restic seems like a better alternative than Arq or Duplicati since I can run it natively on my FreeBSD box..

**Comment by fd0** [comment]
There are no news here, it's one of the more problematic backends, requiring a client secret (that should be kept secret) which needs to be embedded in the Go source code.

You can make a local backup and move that to Google Drive with e.g. rclone.

**Comment by tamalsaha** [comment]
@fd0, from https://developers.google.com/identity/protocols/OAuth2InstalledApp , it seems to me that if this  mode is used the client secret do not need to be kept confidential.

**Comment by naffit** [comment]
@fd0 Alas, that's my initial problem - with 6TB+ of data that is not really an option..

As @tamalsaha mentions others are using OAuth without problems.
The team behind Duplicati has even published this explanation: https://www.duplicati.com/articles/OAuth/ and uses https://duplicati-oauth-handler.appspot.com/ to request/revoke tokens. 

The explanation is based on OneDrive, but it's the same procedure with Google Drive and their OAuth-handler lists a lot of providers based on OAuth.

**Comment by xor-gate** [comment]
Hi guys, I started to play around some time ago with the restic source code to give it a shot to integrate gdrive API package. But there is a lot of duplication going on (needed to change 3 places to add a backend). I have not proceeded because I moved on to some other stuff. https://github.com/xor-gate/restic/compare/master...add-gdrive

**Comment by bherila** [comment]
With Google File Stream you might be able to back up restic today using just the GFS mount point?

**Comment by naffit** [comment]
It works with Drive File Stream - not good or great, but it works. Initializing a repo and running backups on a mounted Drive File Stream drive works. 

Since the Drive File Stream client is implemented with a local cache directory that holds added files _while_ they are uploaded to Google, every restic pack file lands on a local drive before its uploaded (at least when running on Windows), it does introduce some overhead compared to using the Drive API directly.

Due to that I've had to impose upload-limitations when running backups, so my local disk is not attempting to store terabytes of data (that it cant hold) thats not yet uploaded to Google.. 

Somewhere in the Google Product Forums I've found a thread about slow upload speeds that mentions a possible fix due mid or end january, so hopefully it will allow me to keep using Restic and wait for OAuth support.. :)

**Comment by bherila** [comment]
That's a good point. For my own use case, the data to be backed up is significantly smaller than local storage, but I see how it could be a problem if you were backing up a remote or external location using restic.

**Comment by naffit** [comment]
On another note - now that I've actually uploaded close to 260GB over the last few days (damn its slow uploading, so I've opened up a supportcase with Google) I've done a bit of testing.

If there are network dropouts during a backup run I seem to end up with some minor faults that gets detected when I run a check - eg. errors such as:

- pack e4835729fa412ecee86ceabd76ba8a8457de68bfcbbc22c954e91ad50adce8f0: not referenced in any index
- unused blob 67a5b48f
- error for tree 8d322b0d:
   &nbsp;&nbsp;ReadFull(<data/7ceda99e03>): EOF
 
That's easily fixed by running prune and I'm not that worried by it.

However my current "hold your horses" problem is this:

I decided to do run a check with the --read-data option, and while I see my line beeing maxed out now (yeay!) I also see Restic reporting errors such as these:

- Copy: read \\?\G:\My Drive\Restic\Files\data\12\124c44c3e26ef4ee18b7588f31240d8b3756a4c0e7ec2f11b336fd68ff51525f: Incorrect function.
- Copy: read \\?\G:\My Drive\Restic\Files\data\35\3527a45b05e17425eadca8858b27254660030bead68401b79e53833970f738c9: Incorrect function.
- Copy: read \\?\G:\My Drive\Restic\Files\data\38\3892f0c7cf7841e11444a252dfcbf65ecb161c3ca5276337babb21cb1d753393: Incorrect function.

It seems that this is a sideeffect of how Drive File Stream handles requests for readaccess to files and what a possible timeout or "slow request" accessing a file causes.

I've checked my Drive through a browser, and all the problematic files that Restic logs like the above are available (but somehow could not be accessed fast enough).

My current idea is to try and use --limit-download to limit Restic a bit and see if it actually improves things, but right now I'm waiting for the current run of "check --read-data" to complete so I have some sort of baseline for knowing if --limit-download helps.

How far along is pull request #1505 ? I'd be more than happy to test it.. :)
